numcodecs/shuffle.pyx

```diff
@@ -0,0 +1,94 @@
+# cython: embedsignature=True
+# cython: profile=False
+# cython: linetrace=False
+# cython: binding=False
+# cython: language_level=3
+
+import numpy as np
+from .compat import ensure_contiguous_ndarray
+from .abc import Codec
+
+cimport cython
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void _doShuffle(const unsigned char[::1] src, unsigned char[::1] des, Py_ssize_t element_size) nogil:
+    cdef Py_ssize_t count, i, j, offset, byte_index
+    count = len(src) // element_size
+    for i in range(count):
+        offset = i*element_size
+        for byte_index in range(element_size):
+            j = byte_index*count + i
+            des[j] = src[offset + byte_index]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void _doUnshuffle(const unsigned char[::1] src, unsigned char[::1] des, Py_ssize_t element_size) nogil:
+    cdef Py_ssize_t count, i, j, offset, byte_index
+    count = len(src) // element_size
+    for i in range(element_size):
+        offset = i*count
+        for byte_index in range(count):
+            j = byte_index*element_size + i
+            des[j] = src[offset+byte_index]
+
+
+class Shuffle(Codec):
```

Comment:
Would separate the `Shuffle` class into a Python file (say `shuffle.py`) and then leave the Cython bits in `_shuffle.pyx` (underscore as it is internal). This should avoid Cythonizing a class we don't really need Cythonized (and thus avoid any potentially odd behavior that may surface from that). Though would need to mark the function above as `cpdef` to ensure they expose a Python callable API

Reply:
Thanks, have separated out the cython and python as suggested, but on platforms other than linux still getting import failures due to the cython shuffle functions not being recognised. See https://github.com/zarr-developers/numcodecs/pull/273/checks?check_run_id=2113066075 

Not sure how to resolve this other than to wrap the import in the try...catch during the __init__.py